### PR TITLE
update from latest magfest master

### PR DESCRIPTION
**Pull request by binary1230**
ALL THE STUFF, YOU LOVE IT

### CODE REVIEW

**Comment by kitsuta** [comment]
I've been keeping up with the magfest repo, so this gets a +1 from me.

**Comment by binary1230** [comment]
I confess that I did this PR mostly to advocate an approach where all the existing devs (such as myself and Eli) should probably have acecsss to be doing development directly against this repo, with a magfest-specific fork FROM here, instead of risking dealing with large merges back from a magfest-specific repo.  

We're lucky here because there hasn't been any or also any significant activity on this repo.  Had there been any, this merge would have likely been a big nightmare, and un-reviewable since all our features (which are non-magfest-specific and core rams is going to want) are in a giant take-it-or-leave-it-blob.

At the end of the day, the result would have been the same commits moving into this repo, only without a giant merge such as this one involved every couple of months.

**Comment by earl7399** [comment]
No pontification allowed in PRs.  ;)

## 

Robert Earl r.a.earl@gmail.com
FAX: 1-440-919-5086

On Fri, Aug 14, 2015 at 4:40 PM, Dominic Cerquetti <notifications@github.com

**Comment by kitsuta** [comment]
We do definitely want to reverse the relationship of the forks (we need to email GitHub support for this, apparently). That being said, the nearly-constant-emergency-patch workflow of MAGFest does not fit very well with other events. This is the stable version, the PRs are going to take longer (this one is still waiting for review from the other @RAMSProject/owners)... it just doesn't make a lot of sense to try to route updates through here first.

As one example, Eli just had to push through a bunch of rapid-fire fixes for the volunteer checklist, all in a few hours. If he'd had to wait for someone else to review it, we would have missed our deadline - and if he had just merged in his changes to upstream, that would make master an untested (and potentially unstable) branch.

There's a few ways we can have a good compromise, I think. One possibility is that whenever you submit a PR to magfest/master, you'd submit the same one to ramsproject/rams-core - it's a bit of a hassle for developers, but otherwise pretty clean. Another possibility is to just open regular PRs from MAGFest's repo to this one - so instead of a couple months of review to cover, it's only a week, or something. I'm in favor of doing that one for at least right now, so we have a regular workflow of reviewing MAGFest's changes.

**Comment by binary1230** [comment]
I guess what I'm saying is, we could delete this repo, transfer the magfest/ubersystem repo here, and then fork from here to magfest/ubersystem again.

We'd then move all magfest-specific issues out of this repo and into the magfest repo, which would separate out nicely magfest-specific issues from rams generic issues.

If something magfest related is rapid-fire-gotta-have-it-right-now-for-real, then we put it in the magfest repo and try and cherry-pick it back.

If it's not an emergency, then we do the PR against this branch, of which everyone (me vicki eli ksonney and thaeli have access), and that core team is able to judge whether or not the change is appropriate for the generic repo (which everyone there is capable of doing).  So instead of having to wait a while, we have the active dev team actually working on the repo, and we PR the changes from here back to magfest/core

I'm just saying that with very little change in how we're doing things, we can set this up so that this repo isn't behind the magfest repo for 95% of changes, and for the last bit of rapid fire emergency stuff, we can open up PRs via cherry-picking from magfest to here and review/fix those things as we go.

I mean pretty much every single change I and pretty much everyone has made in core lately is going to be wanted here.

The other issue with the current approach is that, this PR is for magfest:master to ramsproject:core, so it in fact contains all the rapid fire fixes in a way that a casual observer isn't going to be able to detect, so they're going to get merged in here.  If @thaeli or @ksonney looks at this PR right now they see Vicki saying '+1' despite the fact that maybe, 12+ commits later from when she made that comment, are commits maybe that shouldn't be +1'd here.

All I'm harping on is the 164 commits from the 3 people currently crushing bugs in ubersystem aren't in this repo, when they could be very easily if we just flip things around a bit.

**Comment by binary1230** [comment]
*If something magfest related is rapid-fire-gotta-have-it-right-now-for-real, then we put it in the magfest repo and try and cherry-pick it back here at everyone's convenience and only perhaps after it's been made less hacky and more worthy of inclusion in the generic rams core.

**Comment by kitsuta** [comment]
Routing stuff through this repo is certainly a possibility that @EliAndrewC brought up. I personally think it would be way too much hassle for everyone involved, but if I'm outvoted on that, it's fine.

Even if we do that though, I really don't believe it's necessary to delete this repo and move the issues by hand - there's almost 200. I have no idea what you gain from doing that, since we already have the commit history here.

Keep in mind that even though including all the code from MAGFest isn't a problem, 80% of the fixes are really just "workflow changes that MAGFest people wanted that are relatively non-offensive to include in core." There's maybe 5 or 10 proper real bugs in here that we're fixing, plus a _bunch_ of pseudo-fixes that will need to be replaced later on.

**Comment by binary1230** [comment]
I have a saying with setting up this kind of stuff which is 'the structure should reflect the reality', and the reality is that we're currently crushing bugs that exist in the core repo, so we should really be working out of this one.  The only real difference is we'd just have to keep in mind that changes here affect other events, which is how we've actually been working for a while now.

For the issues, what I'm saying there is if we did a transfer of ownership of the magfest/ubersystem repo to this organization, it would also transfer all the issues over (which is good as most of them are relevant to everyone using core). But that would catch a few magfest-specific issues which other core users won't care about (like: analyze uber attendance from magfest 10 for demographics).  WIth those issues, we'd close them here, and re-open them in the new magfest fork.  Any new magfest-specific issues would get opened against our repo, and anything that got opened which was in fact non-magfest-specific would get closed there and re-opened here.

Github makes transferring repos painless, so we could do this today if we wanted to.

Also if we did go down this route, I would also suggest we use the /rams/ name for it since I was able to liberate it, and it'll just make the URL shorter.  I think I already added everyone to it.

**Comment by EliAndrewC** [comment]
> As one example, Eli just had to push through a bunch of rapid-fire fixes for the volunteer checklist, all in a few hours. If he'd had to wait for someone else to review it, we would have missed our deadline - and if he had just merged in his changes to upstream, that would make master an untested (and potentially unstable) branch.

For the record though, while I did deliberately merge into MAGFest-specific repos without review (e.g. `production-config`), for the core `ubersystem` repo I did my work in branches and then ssh-ed into the staging and classic servers and manually did a `git pull && git checkout my-feature-branch && supervisorctl restart all`.  Which is not that hard and definitely something we can do going forward so that we can deploy code which isn't merged yet - it's not even the first time I've done that.

I don't have any strong opinions on the merging strategy.  I'd be happy to talk about my views of the pros and cons of each approach, but my guess is that it affects the RAMS folks way more than the MAGFest folks, so I'm happy to give my thoughts and then leave it up to them :)

**Comment by kitsuta** [comment]
> Keep in mind that even though including all the code from MAGFest isn't a problem, 80% of the fixes are really just "workflow changes that MAGFest people wanted that are relatively non-offensive to include in core." There's maybe 5 or 10 proper real bugs in here that we're fixing, plus a bunch of pseudo-fixes that will need to be replaced later on.

What I mean when I say this is that most of the stuff we're doing in the magfest repo is MAGFest-specific. There's no harm in including it in core, but:

> it would also transfer all the issues over (which is good as most of them are relevant to everyone using core)

Is very emphatically _not true._ A huge proportion of the issues in the MAGFest repo are there to change very specific things for the benefit of one department head or another. We don't even know how much other events are going to use the workflows involved in things like the volunteer checklist - I'd like to keep this repo's issues clean so we can focus on the big stuff.

**Comment by ksonney** [comment]
OK, Here are my throughts, although a PR is PROBABLY not the right place to discuss this (someday we will have mailing lists):
- I am in favor of getting the RAMS project name, and the gyrations to make MAG Uber downstream from it. 
- I am concerned that the MAG "fixes" are not all fixes in the traditional sense, but MAG-specific workflow. As we found at AC, the MAG workflow is not universal, and AC had to do a lot of work to "clean up" much of the assumptions MAG makes as to "how things work"
- I would prefer a system where the person who makes the pull request CANNOT merge it. I'd like to see code review and have the person pushing the Merge button be different from the person writing the code.
- One of the things I advocate and WANT to see is more of the things MAG considers "core" (Hotels, Schedules, DH Checklists) broken out to plugins so that they can be enabled/disabled by other events. The same for Dealer-specific enhancements, the pending AC Art show work, and so on. 

Overall, I am inclined to approve this PR, but we need to work out the process for the upstream branch to remain stable.

**Comment by earl7399** [comment]
FWIW, these points reflect my perspective as well.

## 

Robert Earl r.a.earl@gmail.com
FAX: 1-440-919-5086

**Comment by EliAndrewC** [comment]
> One of the things I advocate and WANT to see is more of the things MAG considers "core" (Hotels, Schedules, DH Checklists) broken out to plugins so that they can be enabled/disabled by other events.

I'm pretty sure we're all on the same page about this, in general.  In particular, I'm going to do some hotel work this weekend, and part of that will definitely be splitting that out into its own plugin, since our hotel process is very MAGFest specific.

I'm somewhat surprised to hear schedule and DH checklist on that list, though.  I mean, they would both be fairly easy to split out into their own plugins, so I have no real objection to doing that, but I'd have expected them to stay in the core repo for two main reasons:

1) They both represent things which every event is doing, whether they use RAMS to do it or not.  Every event has an events schedule, though they may use a different (and probably wayyyy nicer) webpage than ours for displaying it.  And every event has department heads (whether they use that title or not) who do things like review which volunteers are coming, etc.

2) Both of those features can just be ignored without affecting anything else.  No emails will be sent out, people won't even be able to see the schedule unless they're given access, the Dept Checklist is empty by default and not even linked to from people's homepages, etc.  So an event which wants to use their own process for those things without involving RAMS can do so with no hassle.

Anyway, just my two cents.

**Comment by binary1230** [comment]
Cool I'm with you @ksonney on those points and think it's reasonable.  

For plugins, in general we're  headed in that direction (notice I implemented the new barcode stuff as a plugin) and putting plugins through the paces with magclassic/prime/AC plugins, analytics is helping make them more robust and easier to write/hook into places.

I do think there's going to be some kind of saturation/ROI point with some features (like what Eli was mentioning about schedules and DH's) that they're going to be "too core" and "reasonable that any event would want this option and could ignore it easily".  i.e. what we're doing is transforming rams from magfest's pile of code into a more generic registration system.  There might be core areas of the system that are very hard to remove and will yield very little additional flexibility if they were removed, so we should judge those carefully.

I think, as always, it's going to take a lot of time and iteration to whittle away core stuff into plugins.  AC is shining a good spotlight on obvious candidates and places where there should be hooks, and other events will do the same, and I think this is going to continue to happen organically.

About not merging your own PR's, I'm with you in spirit there but in reality I'd like to change it slightly from "no one is allowed to MERGE their own PRs" to "no one is allowed to APPROVE their own PRs, but they are responsible for merging after it's approved".    We should define what specifically everyone is comfortable with as far as calling something "approved".  

The reason I'm advocating everyone still merge their own PRs is because of a workflow based around deployments.  If I'm merging in a large PR that everyone approves, but still is large enough that it warrants some special attention to make sure it doesn't break things in production, I want to do that when I'm sitting down in front of a computer with a spare half hour or two to merge, deploy, and test stuff, and have time to fix things.

Let's say someone else merges that code, deploys it, and it breaks unexpectedly in production, all at 3am, I'm the expert on that particular feature and either have to wake up and fix it, or someone else has to spend 2 hours figuring out how to tweak one line of code that would take me 2 seconds to fix.

TLDR: I would advocate a workflow where people can't merge PR's until they're approved, but after that, unless they say otherwise (like 'hey it's cool if anyone merges this'), it's up to them to merge it and babysit the deploy.

Now, here's some consideration for picking an approval method.  Requiring hard approval from all devs all the time for every PR is going to be tough or impossible.  I've been making a huge effort to review all PRs as fast as I get them (as has @kitsuta and @EliAndrewC), and that's been working well.  But, it doesn't always happen all the time, and Eli and Vicki ended up banging on rams with a bunch of PRs that got merged without me while I was busy over the weekend, and I am OK with that.  

My default workflow I advocate for when it's tough for get timely approval from 100% of the parties involved for time-sensitive things is to have it be acceptable for conversations to go like this: "hey guys, here's a feature and unless anyone objects, I'm merging this by +12hours" or similar.  It allows for visibility without bottlenecks.

Basically, it's a tradeoff keeping us more agile and not bottlenecked on crew with very different time schedules, while also having visibility for someone to go "STOP WOAH HEY WAIT, don't merge this I will comment when I get back in front of a computer in a few hours and can explain"

And would also advocate that this workflow should apply in its strictest form to code that is in-production, and should be looser or non-existent for things that are either not in production or experimental, and that everyone should be on the same page about each component's status.

TLDR: we should define our approval process in a way that reflects the reality of everyone's schedules and availability.

Anyway this is all good stuff, let's keep talking.

**Comment by EliAndrewC** [comment]
I agree with most of what Dom said, but I'll make a few comments about this part:

> My default workflow I advocate for when it's tough for get timely approval from 100% of the parties involved for time-sensitive things is to have it be acceptable for conversations to go like this: "hey guys, here's a feature and unless anyone objects, I'm merging this by +12hours" or similar.

In cases where it's important to get something in, MAGFest is going to have its own fork anyway, so the typical workflow will presumably be

```
MAG feature branch --PR---> RAMS master --PR--> MAG master
```

In cases where we need to get something in, and we don't even have time to have anyone review it, we can always just do this

```
MAG feature branch --PR--> MAG master --PR--> RAMS master
```

then we can deploy that way.  Which seems reasonable and isn't even more work for us than doing it the original way.

**Comment by ksonney** [comment]
In the short term, I do like Eli's workflow.

One of the things we do at work - which is part of the model I'm using, 
BTW - is to work it all out on a feature branch,  and test the branch 
before submitting the PR to the stable/production branch. This relies 
on a couple of things in our env :
1. Testing the code before we commit
2. Testing the code on the branch AFTER we commit (i.e. deploy the new 
   branch to a server)
3. Showing the results of the test in the PR

Automated tests and CI would help a bunch on this, and I think working 
towards this is good for the longer term.

## 

-- Kevin Sonney
-- kevin@sonney.com
http://about.me/ksonney
GPG : 1024D/320C 0336 3BC4 13EC 4AEC  6AF2 525F CED7 7BB6 12C9
"quis custodiet ipsos custodes?" (who watches the watchers?)

**Comment by binary1230** [comment]
Ah yea, CI for each feature branch would be amazing.  We'll get there.
